src/components/ImageUpload/ImageUploadProvider.tsx

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2024 Nordeck IT + Consulting GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useWidgetApi } from '@matrix-widget-toolkit/react';
+import { TFunction } from 'i18next';
+import { PropsWithChildren, createContext, useCallback, useRef } from 'react';
+import { ErrorCode, FileRejection } from 'react-dropzone';
+import { useTranslation } from 'react-i18next';
+import { determineImageSize } from '../../lib';
+import { Size } from '../../state';
+import { PromiseSettledResult } from '../../types';
+import { SnackbarDismissAction, SnackbarProps, useSnackbar } from '../Snackbar';
+import { useMaxUploadSize } from './useMaxUploadSize';
+
+/**
+ * Allow a maximum of 32 Megapixels image resolution.
+ * Higher values can lead to rendering performance issues.
+ */
+const maxResolutionPixels = 32_000_000;
+
+/**
+ * Allow a maximum size of 24 MiB per image.
+ * Higher values can lead to rendering performance issues.
+ */
+const maxImageSize = 26_214_400;
+
+type UploadErrors = {
+  /** Name of files that failed due to size restrictions */
+  size: string[];
+  /** Name of files that failed due to resolution restrictions */
+  resolution: string[];
+  /** Name of files that failed for other reasons */
+  other: string[];
+};
+
+type ImageUploadResult = {
+  fileName: string;
+  size: Size;
+  /**
+   * MXC URI pointing to the uploaded image.
+   * {@link https://spec.matrix.org/v1.9/client-server-api/#matrix-content-mxc-uris}
+   */
+  mxc: string;
+};
+
+export type ImageUploadContextState = {
+  /**
+   * Handle a file drop from react-dropzone.
+   * Show a loading snackbar as long as uploads are in progress.
+   * Once all uploads are finished show an error snackbar for errors, if any.
+   * Can be called multiple times while other uploads are still running.
+   *
+   * @param files - Files to be uploaded from react-dropzone
+   * @param rejectedFiles - Files rejected by react-dropzone
+   * @returns Results for the uploaded files
+   */
+  handleDrop: (
+    files: File[],
+    rejectedFiles: FileRejection[],
+  ) => Promise<PromiseSettledResult<ImageUploadResult>[]>;
+  /**
+   * Maximum upload size allowed by the server in bytes.
+   */
+  maxUploadSizeBytes: number;
+};
+
+class ResolutionTooHighError extends Error {
+  constructor() {
+    super('Resolution too high');
+  }
+}
+
+export const ImageUploadContext = createContext<
+  ImageUploadContextState | undefined
```

Comment:
Could be shorter: ImageUploadState
```suggestion
  ImageUploadState | undefined
```

Reply:
Renamed